src/layer/unitary.rs

```diff
@@ -0,0 +1,153 @@
+use num;
+use af;
+use af::{Array, MatProp, Dim4};
+
+use std::sync::{Arc, Mutex};
+use activations;
+use initializations;
+use params::{Input, Params};
+use layer::Layer;
+
+use num::Complex;
+
+
+pub struct Unitary {
+    pub input_size: usize,
+    pub output_size: usize,
+}
+impl Unitary
+{
+    // Wh = (D3 R2 F-1 D2 Pi R1 F D1) h
+    fn wh(params: &Params) -> Array {
+        let t = params.current_unroll;
+        let mut h = params.recurrences[t].clone();
+        let h_size = h.dims()[1] as i64;
+        let h_size_f = h_size as f64;
+ 
+
+        // D1*h
+        let D = af::cplx2(&af::cos(&params.weights[1])
+                                    , &af::sin(&params.weights[1])
+                                    ,false);
+        h = af::mul(&D, &h, true);
+
+
+        // F*
+        h = af::transpose(&af::fft(&af::transpose(&h, false), 1.0, h_size)
+                              , false);
+
+
+        // R1*
+        let h_temp = h.clone();
+        let mut r = af::Array::new(&[num::pow(af::norm(&params.weights[4], af::NormType::VECTOR_2, 1., 1.), 2) as f32]
```

Comment:
Is R1 really a constant? It should be a matrix I believe. Also, fyi: for constants you don't need to cast them to an `af::Array`. You can use them in regular ops, eg: `af::add(&myarray, 3.0f32);`